src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs

```diff
@@ -181,19 +190,31 @@ internal void FinalizeCachingHeaders()
                 // Check if any vary rules exist
                 if (!StringValues.IsNullOrEmpty(varyHeaderValue) || !StringValues.IsNullOrEmpty(varyParamsValue))
                 {
-                    var cachedVaryRules = new CachedVaryRules
+                    // Normalize order and casing of vary by rules
+                    NormalizeStringValues(ref varyHeaderValue);
+                    NormalizeStringValues(ref varyParamsValue);
+
+                    // Update vary rules if they are different
+                    if (State.CachedVaryRules == null ||
+                        !StringValues.Equals(State.CachedVaryRules.VaryRules.Params, varyParamsValue) ||
```

Comment:
These comparisons assume that the stored vary rules are normalized. Assuming that we are the only ones creating these cached entries, this assumption holds.